outputmanager.cpp

```diff
@@ -11,16 +11,59 @@
 
 #include "outputmanager.h"
 
+#include <algorithm>
+#include <iterator>
+
 OutputManager::OutputManager(QObject *parent) : QObject(parent)
 {
 
 }
 
-void OutputManager::handleNewData(const QByteArray &data)
+void OutputManager::operator << (const QByteArray &data)
+{
+    // problem : if a buffer ending with "\r\n" happens to be cut between '\r' and '\n'
+    //           the 2 resulting QString will be interpreted as having 2 line jumps,
+    //           instead of 1, if displayed in a QTextEdit.
+    //           To avoid this, final '\r' are removed from the received buffer,
+    //           and prepended to the next one
+
+    // flag indicating that the previously received buffer ended with CR
+    static bool prev_ends_with_CR = false;
+
+    // append raw data to the buffer, untouched
+    _buffer.append(data);
+
+    QString to_send;
+    if (prev_ends_with_CR)
+    {
+        // CR was removed at the previous buffer, so now we prepend it
+        to_send.append('\r');
+        prev_ends_with_CR = false;
+    }
+
+    if (data.size() > 0)
+    {
+        QByteArray::const_iterator end_cit = data.cend();
+        if (data.endsWith('\r'))
+        {
+            // if buffer ends with CR, we don't copy it
+            end_cit--;
+            prev_ends_with_CR = true;
+        }
+        std::copy(data.begin(), end_cit, std::back_inserter(to_send));
+    }
+
+    // notify that we have new data
+    emit dataConverted(to_send);
+}
```

Comment:
Mi torna quando dici che l'output manager converte "raw data" in dati visualizzabili, ma penso invece che il problema degli `\r` sia una caratteristica di Qt, che riscontri nel QTextEdit. Per esempio se stampi su terminale usando delle printf senza rimuovere gli `\r` mi aspetto che il terminale visualizzi le cose correttamente. 

Per quanto riguarda la rimozione dello `\r`, poiché fondamentalmente lo `\r` è mantenuto da moltissimi anni su Windows solo per compatibilità non credo sia un problema rimuoverlo (per esempio nella visualizzazione di file di test oggi tutti gli editor leggono correttamente un file che contiene solo `\n`). 
In ogni caso, mi sta bene lasciare inalterato quello che ricevi da terminale, ma se il problema degli `\r` è solo un problema di come Qt interpreta i newline (verifica stampando su terminale con i printf), spostiamo il codice nella main window. Ok?